File: go-knowledge-app/internal/shared/shared.go
// Package shared holds low-level helpers — Go-native ports of the Deno app's
// subprocess/filesystem primitives (shared.ts). Intentionally framework-free so
// collectors stay unit-testable in isolation (mirrors the Python _run/read_text
// split).
package shared

import (
	"context"
	"encoding/json"
	"os"
	"os/exec"
	"strings"
	"time"
	"unicode"
)

// RunOptions configures Run. Zero values disable the respective feature.
type RunOptions struct {
	Cwd       string
	TimeoutMs int
}

// Run executes a command and returns its trimmed stdout on success (exit code
// 0), or ("", false) on non-zero exit, missing binary, or timeout. Replaces
// subprocess.run(..., capture_output=True) + the broad except in knowledge.py.
func Run(cmd []string, opts RunOptions) (string, bool) {
	if len(cmd) == 0 {
		return "", false
	}
	ctx := context.Background()
	if opts.TimeoutMs > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(opts.TimeoutMs)*time.Millisecond)
		defer cancel()
	}
	c := exec.CommandContext(ctx, cmd[0], cmd[1:]...)
	if opts.Cwd != "" {
		c.Dir = opts.Cwd
	}
	out, err := c.Output() // captures stdout; stderr discarded like the Deno version
	if err != nil {
		return "", false
	}
	// Trim trailing whitespace only, matching `.replace(/\s+$/, "")`.
	return strings.TrimRightFunc(string(out), unicode.IsSpace), true
}

// ReadText reads a UTF-8 file, returning ("", false) instead of erroring.
func ReadText(path string) (string, bool) {
	b, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	return string(b), true
}

// ParseJSON parses JSON into T, returning (zero, false) on empty/malformed input
// (mirrors a guarded json.loads).
func ParseJSON[T any](text string) (T, bool) {
	var v T
	if text == "" {
		return v, false
	}
	if err := json.Unmarshal([]byte(text), &v); err != nil {
		return v, false
	}
	return v, true
}

File: deno-knowledge-app/src/shared.ts
// Shared low-level helpers — Deno-native replacements for knowledge.py's
// subprocess/filesystem primitives. Intentionally framework-free so collectors
// stay unit-testable in isolation (mirrors the Python `_run`/`read_text` split).

export interface RunOptions {
  cwd?: string;
  timeoutMs?: number;
}

/**
 * Run a command and return its trimmed stdout on success (exit code 0), or
 * `null` on non-zero exit, missing binary, or timeout. Replaces
 * `subprocess.run(..., capture_output=True)` + the broad except in knowledge.py.
 */
export async function run(cmd: string[], opts: RunOptions = {}): Promise<string | null> {
  const [bin, ...args] = cmd;
  try {
    const command = new Deno.Command(bin, {
      args,
      cwd: opts.cwd,
      stdout: "piped",
      stderr: "piped",
      signal: opts.timeoutMs ? AbortSignal.timeout(opts.timeoutMs) : undefined,
    });
    const { code, stdout } = await command.output();
    if (code !== 0) return null;
    return new TextDecoder().decode(stdout).replace(/\s+$/, "");
  } catch {
    return null;
  }
}

/** Read a UTF-8 file, returning `null` instead of throwing on any error. */
export async function readText(path: string): Promise<string | null> {
  try {
    return await Deno.readTextFile(path);
  } catch {
    return null;
  }
}

/** Parse JSON, returning `null` on malformed input (mirrors a guarded json.loads). */
export function parseJson<T = unknown>(text: string | null | undefined): T | null {
  if (!text) return null;
  try {
    return JSON.parse(text) as T;
  } catch {
    return null;
  }
}
